internal/pkg/webconfig/webconfig.go

```diff
@@ -0,0 +1,288 @@
+	}
+	s.mu.Unlock()
+}
+
+// ServeHTTP implements http.Handler.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		s.handleGet(w, r)
+	case "POST":
+		s.handlePost(w, r)
+	default:
+		w.WriteHeader(http.StatusBadRequest)
+	}
+}
+
+// nolint: unparam
+func (s *Server) handleGet(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+
+	_, err := w.Write([]byte(indexPage))
+	if err != nil {
+		s.logger.Printf("webconfig: failed to write index page: %s", err.Error())
+	}
+}
+
+func (s *Server) handlePost(w http.ResponseWriter, r *http.Request) {
+	writeFailure := func(status int, err error) {
+		s.logger.Printf("webconfig: failed to handle POST request: %s", err.Error())
+
+		tmpl, err := template.New("postErr").Parse(configurationRejectedPage)
+		if err != nil {
+			s.logger.Println("failed to parse configuration rejection template:", err)
+			w.WriteHeader(http.StatusInternalServerError)
+
+			return
+		}
+
+		w.WriteHeader(status)
+
+		if err = tmpl.Execute(w, err.Error()); err != nil {
+			s.logger.Println("failed to write response to client:", err)
+		}
+	}
+
+	f, header, err := r.FormFile("config")
+	if err != nil {
+		writeFailure(http.StatusBadRequest, fmt.Errorf("failed to read POST data: %w", err))
+		return
+	}
+
+	if header.Size < 1 {
+		writeFailure(http.StatusBadRequest, fmt.Errorf("invalid header size"))
+
+		return
+	}
+
+	s.configBytes, err = ioutil.ReadAll(f)
+	if err != nil {
+		writeFailure(http.StatusInternalServerError, fmt.Errorf("failed to read file from config POST: %w", err))
+
+		return
+	}
+
+	cfgProvider, err := configloader.NewFromBytes(s.configBytes)
+	if err != nil {
+		writeFailure(http.StatusBadRequest, fmt.Errorf("failed to parse file from POST as a config: %w", err))
+
+		return
+	}
+
+	if err = cfgProvider.Validate(s.mode); err != nil {
+		writeFailure(http.StatusBadRequest, fmt.Errorf("configuration validation failed: %w", err))
+
+		return
+	}
+
+	if _, err = w.Write([]byte(configurationAcceptedPage)); err != nil {
+		writeFailure(http.StatusGone, fmt.Errorf("failed to write response to client: %w", err))
+
+		return
+	}
+
+	// Got a config
+	s.Stop()
+}
+
+// Run executes the configuration receiver, returning any configuration it receives.
+func (s *Server) Run(ctx context.Context) ([]byte, error) {
+	if s.returnSignal == nil {
+		s.returnSignal = make(chan struct{})
+	}
+
+	ips, err := tnet.IPAddrs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of IPs: %w", err)
+	}
+
+	tlsConfig, err := genKeypair(ips)
+	if err != nil {
+		s.logger.Println("failed to generate keypair:", err)
+		return nil, fmt.Errorf("failed to generate self-signed keypair: %w", err)
+	}
+
+	s.svr = &http.Server{
+		Handler:   s,
+		ErrorLog:  s.logger,
+		TLSConfig: tlsConfig,
+	}
+
+	go func() {
+		if err = s.svr.ListenAndServe(); err != nil {
+			s.logger.Println("webconfig: HTTP service stopped:", err)
+		}
+
+		s.Stop()
+	}()
```

Comment:
I know this is already an insecure service, but I wonder if we should do unencypted.

Reply:
Good question.  I had implemented insecure first, and I just didn't remove it when I added the self-signed cert.  Probably just make it a 302 redirect to the https service.

Reply:
:+1: better idea.